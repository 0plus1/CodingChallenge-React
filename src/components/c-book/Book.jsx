import React from 'react';
import BookImage from './BookImage';
import BookDetails from './BookDetails';

function Book (book) {
  return (
    <article className="c-book">
      <div className="c-book__image">
        <BookImage {...book} />
      </div>
      <div className="c-book__details">
        <BookDetails {...book} />
      </div>
    </article>
  );
}

export default Book;
