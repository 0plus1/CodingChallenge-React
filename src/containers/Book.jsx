import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { fetchBooks } from '../actions';
import useFetchBooks from '../hooks/useFetchBooks';
import Layout from '../components/Layout';
import useCurrentBook from '../hooks/useCurrentBook';
import Loading from '../components/Loading';

const BookStyle = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
};

const Book = Layout(({ match }) => {
  const [books] = useFetchBooks();
  const [currentBook] = useCurrentBook(match.params.book_id);

  if (!books
    || books.length === 0
    || !match.params.book_id
    || currentBook === {}
    || currentBook === null
  ) {
    return <Loading />;
  }
  console.log(currentBook);

  return (
    <div data-test="bookComponent" style={BookStyle}>
      {currentBook.book_id}
    </div>
  );
});


Book.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      book_id: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

export default Book;
