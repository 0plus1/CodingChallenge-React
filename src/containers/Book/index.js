import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import BookCard from '../../components/BookCard';
import {
  fetchBookRequest,
} from '../../actions/book';

const Book = ({ bookItems = [], fetchBooks }) => {
  useEffect(() => {
    fetchBooks();
  }, []);

  return (
    <StyledBook>
      <h1>Book Lists</h1>
      <div>
        {
          bookItems.map((book) => (
            <BookCard book={book} key={book.isbn} />
          ))
        }
      </div>
    </StyledBook>
  );
};

const StyledBook = styled.section`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  > div {
    width: 100%;
    max-width: 1024px;
    display: flex;
    flex-wrap: wrap;
  }
`;

const mapStateToProps = (state) => ({
  bookItems: state.book.items,
});

const mapDispatchToProps = (dispatch) => ({
  fetchBooks: () => {
    dispatch(fetchBookRequest());
  },
});

Book.propTypes = {
  bookItems: PropTypes.arrayOf(PropTypes.shape({
    author: PropTypes.string.isRequired,
    book_id: PropTypes.number.isRequired,
    cover: PropTypes.string.isRequired,
    isbn: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    published_at: PropTypes.string.isRequired,
  })).isRequired,
  fetchBooks: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(Book);
