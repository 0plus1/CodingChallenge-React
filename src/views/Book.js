import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import BookResult from '../components/BookResult';
import Layout from './Layout';
import { connect } from 'react-redux';
import { loadBooks } from '../actions/bookActions';
import Loading from './Loading';

const BookView = ({ book, books, dispatch }) => {

  // if the route is accessed directly there will be no book to find - so go fetch
  useEffect(() => {
    if (books.length < 1) {
      dispatch(loadBooks());
    }
  }, [dispatch, books]);

  const render = book ? <BookResult {...book} /> : <Loading />;
  return render;
};

BookView.propTypes = {
  book: PropTypes.shape(BookResult.propTypes),
};

const mapStateToProps = (state, ownProps) => {
  const { bookId } = ownProps.match.params;
  const foundBook = state.books.find(book => book.bookId === parseInt(bookId, 10));

  let book = foundBook ? { ...foundBook } : null;

  return {
    book,
    books: state.books
  };
};

export default Layout(connect(mapStateToProps)(BookView));
