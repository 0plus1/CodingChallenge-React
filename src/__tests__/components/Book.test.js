import React from 'react';
import { shallow } from 'enzyme';

import Book from '../../components/Book';

const book = {
    "book_id": 1,
    "name": "Becker West Arnoldo",
    "isbn": "3181444340",
    "published_at": "2000-01-01",
    "author": "Mrs. John Doe",
    "cover": "https:\/\/lorempixel.com\/640\/480\/?82539"
};

describe('Book', () => {
    it('should render correctly', () => {
        const bookList = shallow(<Book name={book.name} author={book.author} />);
        expect(bookList).toMatchSnapshot();
    });
});

