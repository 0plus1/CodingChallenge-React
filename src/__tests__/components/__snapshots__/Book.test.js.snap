// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book should render correctly 1`] = `ShallowWrapper {}`;
